src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -44138,6 +44138,131 @@ static void Main()
                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "F<IOut<object>>").WithArguments("t", "void C.F<IOut<object>>(ref IOut<object> t)", "D<IOut<object?>>").WithLocation(17, 34));
         }
 
+        [Fact]
+        [WorkItem(37984, "https://github.com/dotnet/roslyn/issues/37984")]
+        public void DelegateCreation_FromCompatibleDelegate()
+        {
+            var source = @"
+using System;
+
+delegate void D();
+
+class C
+{
+    void M(Action a1, Action? a2, D d1, D? d2)
+    {
+        _ = new D(a1);
+        _ = new D(a2); // 1
+        _ = new D(a2!);
+        _ = new Action(d1);
+        _ = new Action(d2); // 2
+        _ = new Action(d2!);
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (11,19): warning CS8601: Possible null reference assignment.
+                //         _ = new D(a2); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "a2").WithLocation(11, 19),
+                // (14,24): warning CS8601: Possible null reference assignment.
+                //         _ = new Action(d2); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "d2").WithLocation(14, 24));
+        }
+
+        [Fact]
+        [WorkItem(37984, "https://github.com/dotnet/roslyn/issues/37984")]
+        public void DelegateCreation_FromNullableMismatchedDelegate()
+        {
+            var source = @"
+using System;
+
+delegate void D1(string s);
+delegate void D2(string? s);
+
+class C
+{
+    void M(Action<string> a1, Action<string?> a2, D1 d1, D2 d2)
+    {
+        _ = new D1(a1);
+        _ = new D1(a2);
+        _ = new D2(a1); // 1
+        _ = new D2(a1!);
+        _ = new D2(a2);
+        _ = new Action<string>(d1);
+        _ = new Action<string>(d2);
+        _ = new Action<string?>(d1); // 2
+        _ = new Action<string?>(d1!);
+        _ = new Action<string?>(d2);
+    }
```

Comment:
> } [](start = 3, length = 2)

Consider testing:
```
_ = new D(d1);
_ = new D(d2);
_ = new D(default(D));
```

Reply:
Please test nested nullability mismatch:
```C#
delegate void D<T>(T t);
class Program
{
    static void M(D<object> d)
    {
        _ = new D<object?>(d);
    }
}
```

---
In reply to: [315376466](https://github.com/dotnet/roslyn/pull/38063#discussion_r315376466) [](ancestors = 315376466)